pipeline/id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/steps/IdentifierGenerator.scala

```diff
@@ -1,41 +1,48 @@
 package uk.ac.wellcome.platform.idminter.steps
 
+import grizzled.slf4j.Logging
 import uk.ac.wellcome.models.work.internal.SourceIdentifier
 import uk.ac.wellcome.platform.idminter.database.IdentifiersDao
 import uk.ac.wellcome.platform.idminter.models.Identifier
 import uk.ac.wellcome.platform.idminter.utils.Identifiable
 
-import scala.util.Try
+import scala.collection.breakOut
+import scala.util.{Success, Try}
 
-class IdentifierGenerator(identifiersDao: IdentifiersDao) {
-
-  def retrieveOrGenerateCanonicalId(
-    identifier: SourceIdentifier
-  ): Try[String] =
-    Try {
-      identifiersDao
-        .lookupId(
-          sourceIdentifier = identifier
-        )
-        .flatMap {
-          case Some(id) => Try(id.CanonicalId)
-          case None     => generateAndSaveCanonicalId(identifier)
-        }
-    }.flatten
-
-  private def generateAndSaveCanonicalId(
-    sourceIdentifier: SourceIdentifier
-  ): Try[String] = {
-    val canonicalId = Identifiable.generate
+class IdentifierGenerator(identifiersDao: IdentifiersDao) extends Logging {
+  // This function fetches canonicalIds for sourceIdentifiers, and generates
+  // and saves canonicalIds where it can't find existing ones. For this reason
+  // it's important not to run it in situations where there might be a race
+  // condition (IDs being inserted in between lookup and update).
```

Comment:
What can I do to prevent race conditions?

Isn't that what MySQL integrity checks are for? If two processes try to create and save identifiers at the same time, only one of them will win, the other will throw an exception and the message will either retry or be dropped from the queue.

Reply:
That is true - whilst I believe it is still technically a race condition, it would cause issues 100% of the time. I'll try to explain...

Consider `n` messages containing identical and unminted `sourceIdentifiers` arriving in the minter at the same time.

Because we are good developers, we know we should be able mint everything concurrently, so we use a `mapAsyncUnordered` on the message stream, creating `n` threads - one for each message. Let's simplify and say we only have one processing core, so they'll never _actually_ be running at the same time.

JDBC calls are inherently blocking, so when a thread gets to the `lookup` operation it will block and the scheduler will switch to one of the other `n - 1` threads, where the same thing will happen, and before we know it all `n` threads are doing their lookup at once and finding nothing for the `sourceIdentifier`.

Each thread will then generate a new identifier and try to save it. One of them will get there first, and the other `n - 1` will fail with certainty and need to be redriven, every time! 😞